### ISSUE TITLE
report errors when attempting to create torrents with 0 bytes payload

### ISSUE THREAD

**Issue by aldenml** [arvidn/libtorrent#637]
Hi, before starting to write any code, I would like to hear from you regarding the possibility of this case. I added the restriction to `set_piece_hashes` because having not space in the inner vector `m_piece_hash` becomes a crash. But I can't clearly see if this is an artificial restriction or there is no room for this case in the torrent concept.

**Comment by arvidn** [comment]
I don't think torrents with only 0-sized files are valid. i.e. I think it's fine to disallow them

**Comment by aldenml** [comment]
Not mathematical proof here :)

**Comment by Col-blimp** [comment]
I hope you can change this to accept zero bit files as other clients do accept them so its just potentially splitting swarms.

**Comment by arvidn** [comment]
@Col-blimp can you be more specific about what kinds of torrents are currently not accepted that you think should be? I am under the impression that other clients also disallow torrents of size 0, regardless of the number of files

**Comment by Col-blimp** [comment]
Hi there, utorrent accepts zero bit files if the folder has other file(s) that are greater than zero bytes, 

I think your right and its a redundant exercise but it does split the swarm if a torrent created is a duplicate,

**Comment by arvidn** [comment]
it sounds like you mean torrents that are greater than zero sized though, which are supported. I don't understand the splitting of swarms. the info-hash would be identical regardless, no?

**Comment by Col-blimp** [comment]
The info hash is different as there would be an additional file in the torrent albeit of zero size

Example:

folder: text doc zero bytes, text doc with 'hello' = DCC4429B7651C62D44CC85D5C1E55A0A000E1898
folder text doc with 'hello' = A7921CB99DB46254DD2B9045CD9A9D12A288F141

**Comment by arvidn** [comment]
are you talking about *creating* torrents? i.e. the create_torrent class in libtorrent. If you're interested in *not* splitting swarms, I assume you would like libtorrent to *ignore* empty files (when creating torrents). that way anyone adding empty files would not affect the info-hash, but presumably that user would also be disappointed that libtorrent ignored that file, and perhaps just use another torrent creator. My understanding is that all relevant torrent creators *do* include empty files, because that's a feature people want.

**Comment by Col-blimp** [comment]
Yes, I joined this topic as someone refereed me to it on the mailing-list as create_torrent when a folder containing zero bit files is created into a torrent it creates an invalid torrent file or simply crashes if the torrent is a single file of zero bytes, I've filtered out all zero bit files anyhow now but then thought it might be better having them in torrents created from folders, sorry if this is the wrong place for this.

**Comment by arvidn** [comment]
I'm not entirely sure what @aldenml had in mind for the topic of this ticket, but I believe it was related to https://github.com/arvidn/libtorrent/pull/642

@Col-blimp If you have a case where libtorrent creates invalid torrents, please create a new ticket and describe the steps to reproduce it as well as version of libtorrent and platform.

**Comment by aldenml** [comment]
I want to clarify that https://github.com/arvidn/libtorrent/pull/642 just add a few asserts, but in a release build, a crash from a bad usage is not prevented (manual setup of the piece hashes). The only protection was added in https://github.com/arvidn/libtorrent/pull/636/files for the function `set_piece_hashes`.

**Comment by arvidn** [comment]
ah, I see. it would probably make sense to have generate() fail. really the create_torrent constructor too, but I might want to ease-in pervasive use of exceptions (to allow constructors to fail)
